src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -1479,6 +1470,188 @@ static void Main()
+        System.Console.Write(C1.M().Item1 + "" "");
+        System.Console.Write(C1.M().a + "" "");
+        System.Console.Write(C1.M().Item2 + "" "");
+        System.Console.Write(C1.M().b + "" "");
+        System.Console.Write(C2.M().Item1 + "" "");
+        System.Console.Write(C2.M().c + "" "");
+        System.Console.Write(C2.M().Item2 + "" "");
+        System.Console.Write(C2.M().d);
+    }
+}
+";
+            var comp1 = CreateCompilationWithMscorlib(source1);
+            var comp2 = CreateCompilationWithMscorlib(source2);
+            var comp = CompileAndVerify(source, expectedOutput: @"1 1 2 2 3 3 4 4", additionalRefs: new[] { new CSharpCompilationReference(comp1), new CSharpCompilationReference(comp2) });
+        }
+
+        [Fact]
+        public void DistinctTupleTypesInCompilationCannotAssign()
+        {
+            var source1 = @"
+public class C1
+{
+    public static (int a, int b) M()
+    {
+        return (1, 2);
+    }
+}
+" + trivial2uple;
+
+            var source2 = @"
+public class C2
+{
+    public static (int c, int d) M()
+    {
+        return (3, 4);
+    }
+}
+" + trivial2uple;
+
+            var source = @"
+class C3
+{
+    public static void Main()
+    {
+        var x = C1.M();
+        x = C2.M();
+    }
+}
+";
+            var comp1 = CreateCompilationWithMscorlib(source1);
+            var comp2 = CreateCompilationWithMscorlib(source2);
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { new CSharpCompilationReference(comp1), new CSharpCompilationReference(comp2) });
+
+            // PROTOTYPE(tuples) this error is misleading or worse.
+            comp.VerifyDiagnostics(
+                // (7,13): error CS0029: Cannot implicitly convert type '<tuple: int c, int d>' to '<tuple: int a, int b>'
+                //         x = C2.M();
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "C2.M()").WithArguments("<tuple: int c, int d>", "<tuple: int a, int b>").WithLocation(7, 13)
+                );
+        }
+
+        [Fact]
+        public void AmbiguousTupleTypesForCreation()
+        {
+            var source = @"
+class C3
+{
+    public static void Main()
+    {
+        var x = (1, 1);
+    }
+}
+";
+            var comp1 = CreateCompilationWithMscorlib(trivial2uple);
+            var comp2 = CreateCompilationWithMscorlib(trivial2uple);
+
+            var comp = CompileAndVerify(source, additionalRefs: new[] { new CSharpCompilationReference(comp1), new CSharpCompilationReference(comp2) });
+            comp.VerifyDiagnostics(
+                // warning CS1685: The predefined type 'ValueTuple<T1, T2>' is defined in multiple assemblies in the global alias; using definition from '__assembly_guid__, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
+                Diagnostic(ErrorCode.WRN_MultiplePredefTypes).WithArguments("System.ValueTuple<T1, T2>", comp1.AssemblyName + ", Version=0.0.0.0, Culture=neutral, PublicKeyToken=null").WithLocation(1, 1)
```

Comment:
I had forgotten to compose the expected diagnostics with `comp.AssemblyName` in one of the tests. But I like your suggestion better. I'm setting the assemblyName on the compilation now.